lib/burnside/convert-expr.cc

```diff
@@ -714,6 +696,38 @@ class ExprLowering {
     return std::visit([&](const auto &e) { return genval(e); }, exp.u);
   }
 
+  template<int KIND>
+  M::Value *genval(Ev::Expr<Ev::Type<LogicalCat, KIND>> const &exp) {
+    auto *rawResult{
+        std::visit([&](const auto &e) { return genval(e); }, exp.u)};
+    // Handle the `i1` to `fir.logical` conversions as needed.
+    auto *typedResult{rawResult};
+    if (rawResult) {
+      M::Type type{rawResult->getType()};
+      if (type.isa<fir::LogicalType>()) {
+        if (genLogicalAsI1) {
+          M::Type mlirLogicalType{getMLIRlogicalType(builder.getContext())};
+          typedResult = builder.create<fir::ConvertOp>(
+              getLoc(), mlirLogicalType, rawResult);
```

Comment:
Yes it is forecasting how logical will be used. The implementation is a "Propagate as consumed" (as used) which is different from convert on `use`. I tried to described this in the design doc, I will try to rephrase it here:

The point is mainly to centralize the spot where conversions are inserted and to provide a defined behavior for code that would lower logical expression without handling the  MLIR type alternatives it could get for the lowered expression. If the code does not explicitly ask for `i1` it will get a `fir.logical`.

Part of the issue of delaying conversions until the actual consumer is that in brings uncertainty in all code that is propagating the value from the code that inserted the producer until the code that is inserting a consumer. For instance, this brings issues around operations that require operands of the same type but otherwise do not care about the actual type.

Also, "forecasting" is actually more keeping track of a requirement rather than looking ahead for it. While lowering, we go down the parse and expression trees, so we go from consumer to producer and it is not that hard to keep track of the consumer logical type requirements.
 
So this the logic is not exactly a convert on use, but a "user ask what it wants and if it does not ask, it gets fir.logical". IMHO, it is more future proof. If you disagree here or this is not clear, please comment inside the related design doc so that I can fix it there.